components/builder-worker/src/runner/util.rs

```diff
@@ -0,0 +1,216 @@
+            }
+        }
+        // Optional keys with string values
+        if let Some(val) = opts.get("custom_tag") {
+            if !val.is_string() {
+                return Err(Error::InvalidIntegrations(format!(
+                    "project integration custom_tag value must be a string"
+                )));
+            }
+        }
+    }
+    // Validate origin integration
+    {
+        let org_integrations = workspace.job.get_integrations();
+        if org_integrations.is_empty() {
+            return Err(Error::InvalidIntegrations(format!(
+                "missing Docker credentials from origin integrations"
+            )));
+        }
+        let org_integration = org_integrations.first().unwrap();
+        if org_integration.get_integration() != "docker" {
+            return Err(Error::InvalidIntegrations(format!(
+                "origin integration '{}' not supported",
+                org_integration.get_integration()
+            )));
+        }
+        if org_integration.get_name() != "docker" {
+            return Err(Error::InvalidIntegrations(format!(
+                "origin integration name '{}' not supported",
+                org_integration.get_name()
+            )));
+        }
+        // TODO fn: use a struct and serde to do heavy lifting
+        let creds: JsonValue = match serde_json::from_str(org_integration.get_body()) {
+            Ok(json) => json,
+            Err(err) => {
+                return Err(Error::InvalidIntegrations(format!(
+                    "origin integration body does not deserialize as JSON: {:?}",
+                    err
+                )))
+            }
+        };
+        // Required keys with string values
+        for str_key in vec!["username", "password"].iter() {
+            match creds.get(str_key) {
+                Some(s) => {
+                    if !s.is_string() {
+                        return Err(Error::InvalidIntegrations(format!(
+                            "origin integration {} value must be a string",
+                            str_key
+                        )));
+                    }
+                }
+                None => {
+                    return Err(Error::InvalidIntegrations(
+                        format!("origin integration {} missing", str_key),
+                    ));
+                }
+            }
+        }
+    }
+    debug!("validated integrations");
+    Ok(())
+}
+
+
+/// Builds the Docker exporter details from the origin and project integrations.
+pub fn docker_exporter_spec(workspace: &Workspace) -> DockerExporterSpec {
+    // TODO fn: Using this value struct was done to keep the validation, JSON parsing, and
+    // craziness in one place and out of the Docker-specific code. Oi, this is embarrassing.
+    let creds: JsonValue = serde_json::from_str(
+        workspace
+            .job
+            .get_integrations()
+            .first()
+            .expect("Origin integrations must not be empty")
```

Comment:
Is a panic what we want here?

Reply:
It's not ideal, but at this stage, that data structure has been validated and hasn't been modified. I was just loathe to do the same error reporting back. Think it's worth it?

Reply:
If it's already been validated, then that's cool. Maybe add some comments to that effect? Likely not worth adding in extra error reporting at this point, since it's clear that all this is going to get massively refactored to use `serde` & friends real soon anyway.

![tenor-36183838](https://user-images.githubusercontent.com/207178/31098405-6b271046-a790-11e7-92e3-75c3949ffa81.gif)

Ship it! 😄

Reply:
Sounds good, a comment is well worth it: `.unwrap()` and `.expect()` should usually be explained in our codebase anyway.

Reply:
Comment added, good call!